clang/lib/Sema/SemaSYCL.cpp

```diff
@@ -1822,6 +1838,90 @@ static std::string getCPPTypeString(QualType Ty) {
   return eraseAnonNamespace(Ty.getAsString(P));
 }
 
+static void printArguments(ASTContext &Ctx, raw_ostream &ArgOS,
+                           ArrayRef<TemplateArgument> Args,
+                           const PrintingPolicy &P, bool ParameterPack) {
+
+  if (!ParameterPack)
```

Comment:
Why does the packness matter?

Reply:
We don't want '<' printed each time we print an argument if its part of a pack.

Reply:
Ah, I see now... Between that and the commas, this function seems pretty disorganized.  It seems to me this needs to be split into a 'handle single arg' and a 'loop through things' version (and perhaps 1 to inroduce the top level angle brackets).  This function is attempting to do too much.

Reply:
My first attempt at this function actually didn't handle parameter pack correctly and when I was implementing the fix for that I came across printTo( ) in TypePrinter.cpp which looked like it was handling what I wanted to and so I just incorporated that code into this function. 
https://clang.llvm.org/doxygen/TypePrinter_8cpp_source.html#l01755

Could you elaborate on what you mean by separating it out? The loop is going through every argument in an argument list individually and printing it. If it sees a paramater pack it calls the function with the pack as the argument list (and goes through each argument of the pack individually).

Reply:
something like:

printTemplateArguments
    os << "<";
    printArguments
    os << ">".

printArguments
   for-each-ag
       printArgument
       os << "," // obviously skip the last one

printArgument
   switch(argtype)
       case <whatever>: ....
       case Pack:
          printArguments(Pack.getArray);

Reply:
the spacing on each of those got messed up... 3 functions, just split up the behavior a bunch.

Reply:
Ok. Will do. Thanks!

Reply:
Done